pages/manager/EmailHandler.inc.php

```diff
@@ -211,6 +211,197 @@ function enableEmail($args) {
+		
+		$uploadName = 'email_file';
+		$fileName = $fileManager->getUploadedFileName($uploadName);
+		if (!$fileName) {
+			$request->redirect(null, null, 'emails');
+		}
+		
+		$filesDir = Config::getVar('files', 'files_dir');
+		$filePath = $filesDir . '/journals/' . $journalId . '/' . $fileName;
+		
+		if (!$fileManager->uploadError($uploadName)) {
+			if ($fileManager->uploadedFileExists($uploadName)) {
+				if ($fileManager->uploadFile($uploadName, $filePath)) {
+					if ($this->_isValidXML($filePath)) {
+						if ($this->_saveEmailTemplates($filePath, $journal)) {
+							if ($fileManager->deleteFile($filePath)) {
+								$this->_showMessage($request);
+								$request->redirect(null, null, 'emails');
+								return true;
+							}
+						}
+					}
+				}
+			}
+		}
+		
+		$this->_showMessage($request, false);
+		$request->redirect(null, null, 'emails');
+		return false;
+	}
+	
+	/**
+	 * Save a custom email template file
+	 * @param $filePath string
+	 * @param $journalId int
+	 * @return boolean
+	 */
+	function _saveEmailTemplates($filePath, $journal) {
+		import('lib.pkp.classes.xml.XMLParser');
+		$emailTemplateDao = DAORegistry::getDAO('EmailTemplateDAO');
+		
+		$xmlParser = new XMLParser();
+
+		$struct = $xmlParser->parseStruct($filePath);
+		$locale = $struct['email_texts'][0]['attributes']['locale'];
+
+		$emailTexts = $struct['email_text'];
+		$subjects = $struct['subject'];
+		$bodies = $struct['body'];
+		
+		$journalId = $journal->getId();
+		$supportedLocales = $journal->getSupportedLocaleNames();
+
+		foreach($emailTexts as $index => $emailText) {
+			$emailKey = $emailText['attributes']['key'];
+			$subject = $subjects[$index]['value'];
+			$body = $bodies[$index]['value'];
+
+			$emailTemplate = $emailTemplateDao->getLocaleEmailTemplate($emailKey, $journalId);
+			$emailTemplateLocaleData = $emailTemplate->localeData;
+			
+			// just update supported locales
+			foreach($emailTemplateLocaleData as $emailTemplateLocale => $data) {
+				if (!isset($supportedLocales[$emailTemplateLocale])) {
+					unset($emailTemplateLocaleData[$emailTemplateLocale]);
+				}
+			}
+			$emailTemplate->localeData = $emailTemplateLocaleData; 
+			
+			$emailTemplate->setAssocType(ASSOC_TYPE_JOURNAL);
+			$emailTemplate->setAssocId($journalId);
+			
+			if ($emailTemplate->getCanDisable()) {
+				$emailTemplate->setEnabled($emailTemplate->getData('enabled'));
+			}
+			
+			$emailTemplate->setSubject($locale, $subject);
+			$emailTemplate->setBody($locale, $body);
+
+			if ($emailTemplate->getEmailId() != null) {
+				$emailTemplateDao->updateLocaleEmailTemplate($emailTemplate);
+			} else {
+				$emailTemplateDao->insertLocaleEmailTemplate($emailTemplate);
+			}
+		}
+		return true;
+	}
+	
+	/**
+	 * Check if an XML file contains valid XML
+	 * @param $filePath string
+	 * @return boolean
+	 */
+	function _isValidXML($filePath) {
+		if (is_file($filePath)) {
+			return is_object(simplexml_load_file($filePath));
```

Comment:
This results in double-parsing the file, and adds a simplexml dependency we don't have elsewhere.